### PR TITLE
Rest api multiselect attribute bc break between 2.1.8 and 2.3.0

### PR DESCRIPTION
**Pull request by Valant13**
### Description (*)

Correct conditions for multiselect attribute validation have been added.
Unit test for app/code/Magento/Eav/Model/TypeLocator/SimpleType.php has been created.

### Fixed Issues (if relevant)

1. magento/magento2#21364: Rest api multiselect attribute bc break between 2.1.8 and 2.3.0

### Manual testing scenarios (*)

1. Create multiple select product attribute
2. Create product attribute set
3. Create simple product and assign it to created attribute set
4. Update product using PUT method of catalogProductRepositoryV1 API

Sample JSON body (options as string with ids separated by comma):
```
{
  "product": {
    "sku": "Test Simple Product",
    "custom_attributes": [
      {
        "attribute_code": "test_multiselect",
        "value": "4,5"
      }
    ]
  },
  "saveOptions": true
}
```

Sample JSON body (options as array of ids in string):
```
{
  "product": {
    "sku": "Test Simple Product",
    "custom_attributes": [
      {
        "attribute_code": "test_multiselect",
        "value": ["4", "5"]
      }
    ]
  },
  "saveOptions": true
}
```

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=21901) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @Valant13. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/Eav/Model/TypeLocator/SimpleType.php [hunk 1]
```diff
@@ -50,6 +50,13 @@ public function getType($attributeCode, $entityType)
         } catch (NoSuchEntityException $exception) {
             return TypeProcessor::NORMALIZED_ANY_TYPE;
         }
+
+        $arrayFrontendInputs = ['multiselect'];
+        $frontendInput = $attribute->getFrontendInput();
+        if (in_array($frontendInput, $arrayFrontendInputs)) {
```
Comment (orlangur) [inline app/code/Magento/Eav/Model/TypeLocator/SimpleType.php]:
```suggestion
        if ($attribute->getFrontendInput() === 'multiselect') {
```

File: app/code/Magento/Eav/Test/Unit/Model/TypeLocator/SimpleTypeTest.php [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+     * @param string $getBackendTypeReturn
+     * @param string $expected
+     * @dataProvider getTypeDataProvider
+     */
+    public function testGetType($serviceTypeListMock, $getFrontendInputReturn, $getBackendTypeReturn, $expected)
+    {
+        /**
+         * @var Attribute|\PHPUnit_Framework_MockObject_MockObject
+         */
+        $attributeMock = $this->getMockBuilder(Attribute::class)
```
Comment (orlangur) [inline app/code/Magento/Eav/Test/Unit/Model/TypeLocator/SimpleTypeTest.php]: Please fix formatting and use `createMock` instead to make mock declarations more compact.

File: app/code/Magento/Eav/Test/Unit/Model/TypeLocator/SimpleTypeTest.php [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                                              ->setMethods(['get'])
+                                              ->getMock();
+        $attributeRepositoryMock->expects($this->any())
+                                      ->method('get')
+                                      ->willReturn($attributeMock);
+
+        /**
+         * @var SimpleType|\PHPUnit_Framework_MockObject_MockObject
+         */
+        $simpleType = ($this->objectManger)->getObject(
```
Comment (orlangur) [inline app/code/Magento/Eav/Test/Unit/Model/TypeLocator/SimpleTypeTest.php]: Get rid of ObjectManager and `setUp` method.

File: app/code/Magento/Eav/Test/Unit/Model/TypeLocator/SimpleTypeTest.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        );
+
+        $this->assertSame($expected, $simpleType->getType('testAttributeCode', 'testEntityType'));
+    }
+
+    /**
+     * @return array
+     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
+     */
+    public function getTypeDataProvider()
```
Comment (orlangur) [inline app/code/Magento/Eav/Test/Unit/Model/TypeLocator/SimpleTypeTest.php]: Such test does not actually verify anything. It's just a method logic written in a different form.

Try splitting into proper test cases.

**Review by orlangur — CHANGES_REQUESTED** [review]
Thanks for contribution @Valant13!

Please check my comments, simplify implementation and unit test.

Also, please write an API test covering initial case and after all builds are green, squash changes into a single commit.

File: app/code/Magento/Eav/Model/TypeLocator/SimpleType.php (hunk 1 above)
Comment (Valant13) [reply app/code/Magento/Eav/Model/TypeLocator/SimpleType.php]: Hi, thank you for the review.
Do you mean integration tests?

Comment (orlangur) [reply app/code/Magento/Eav/Model/TypeLocator/SimpleType.php]: Nope, api-functional: https://github.com/magento/magento2/tree/2.3-develop/dev/tests/api-functional/testsuite/Magento/Eav/Api

**Comment by sidolov** [comment]
Hi @Valant13  , looks like you made some commits with email different than in your GitHub profile, please, add email from commits to your profile!
Thank you!

File: app/code/Magento/Eav/Test/Unit/Model/TypeLocator/SimpleTypeTest.php (hunk 3 above)
Comment (Valant13) [reply app/code/Magento/Eav/Test/Unit/Model/TypeLocator/SimpleTypeTest.php]: Hi @orlangur. I use ObjectManager to create testable class instance. What should I use instead ?

Comment (orlangur) [reply app/code/Magento/Eav/Test/Unit/Model/TypeLocator/SimpleTypeTest.php]: I meant just get rid of property and `setUp` method but here it also makes sense to replace it with `new` operator.

**Review by orlangur — CHANGES_REQUESTED** [review]

**Review by orlangur — CHANGES_REQUESTED** [review]
@Valant13 perfect! 👍 

LGTM, please squash changes into a single commit so that we have perfectly clean history 😉

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-4920](https://jira.corp.magento.com/browse/ENGCOM-4920) has been created to process this Pull Request

**Review by orlangur — CHANGES_REQUESTED** [review]
Changes are still not squashed. @Valant13 please do it.

**Comment by Valant13** [comment]
@orlangur, sorry for delay. I will squash changes on the next week

**Comment by Valant13** [comment]
@orlangur, I have squashed all changes to the single commit

**Review by orlangur — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @orlangur, thank you for the review.
[ENGCOM-4920](https://jira.corp.magento.com/browse/ENGCOM-4920) has been created to process this Pull Request

**Comment by soleksii** [comment]
:heavy_check_mark: QA Passed

**Before:**

before

**After:**

after

**Comment by Nazar65** [comment]
Hi @Valant13 thank you for collaboration, yours changes  doesn't work for SOAP client, and  test testUpdateMultiselectAttributesWithArrayValue() failed with following result
```
SoapFault : Notice: Array to string conversion in /home/nazar/sites/magento/publicPr/22/app/code/Magento/Eav/Model/Entity/Attribute/Backend/ArrayBackend.php on line 47
 /home/nazar/sites/magento/publicPr/22/vendor/zendframework/zend-soap/src/Client.php:1166
 /home/nazar/sites/magento/publicPr/22/dev/tests/api-functional/framework/Magento/TestFramework/TestCase/Webapi/Adapter/Soap.php:55
 /home/nazar/sites/magento/publicPr/22/dev/tests/api-functional/framework/Magento/TestFramework/TestCase/WebapiAbstract.php:176
 /home/nazar/sites/magento/publicPr/22/dev/tests/api-functional/testsuite/Magento/Catalog/Api/ProductRepositoryInterfaceTest.php:1236
 /home/nazar/sites/magento/publicPr/22/dev/tests/api-functional/testsuite/Magento/Catalog/Api/ProductRepositoryInterfaceTest.php:1620
```

**Comment by VladimirZaets** [comment]
@Valant13, I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for collaboration

**Pull request closed without merging**